genvm/rewards.go

```diff
@@ -2,79 +2,107 @@ package vm
+	)
+	for _, blockReward := range blockRewards {
+		totalWeight.Add(util.WeightFromNumDenom(blockReward.Weight.Num, blockReward.Weight.Denom))
 	}
-	finalRewards := make([]*types.Reward, 0, len(rewards))
-	layerRewards := calculateLayerReward(cfg)
-	totalRewards := layerRewards + totalFees
-	logger.With().Debug("rewards info for layer",
-		log.Uint64("layer_rewards", layerRewards),
-		log.Uint64("fee", totalFees))
-	rewardPer := util.WeightFromUint64(totalRewards).Div(totalWeight)
-	lyrRewardPer := util.WeightFromUint64(layerRewards).Div(totalWeight)
-	seen := make(map[types.Address]struct{})
-	for _, reward := range rewards {
-		if _, ok := seen[reward.Coinbase]; ok {
-			continue
+	for _, blockReward := range blockRewards {
+		relative := util.WeightFromNumDenom(blockReward.Weight.Num, blockReward.Weight.Denom).Div(totalWeight)
+
+		totalReward := new(big.Int).SetUint64(total)
+		totalReward.
+			Mul(totalReward, relative.Num()).
+			Quo(totalReward, relative.Denom())
+		if !totalReward.IsUint64() {
+			return fmt.Errorf("%w: total reward %v for %v overflows uint64",
+				core.ErrInternal, totalReward, blockReward.Coinbase)
+		}
+
+		subsidyReward := new(big.Int).SetUint64(subsidy)
+		subsidyReward.
+			Mul(subsidyReward, relative.Num()).
+			Quo(subsidyReward, relative.Denom())
+		if !subsidyReward.IsUint64() {
+			return fmt.Errorf("%w: subsidy reward %v for %v overflows uint64",
+				core.ErrInternal, subsidyReward, blockReward.Coinbase)
 		}
 
-		seen[reward.Coinbase] = struct{}{}
-		weight := byCoinbase[reward.Coinbase]
+		v.logger.With().Debug("rewards for coinbase",
+			lctx.Layer,
+			blockReward.Coinbase,
+			log.Stringer("relative weight", &blockReward.Weight),
+			log.Uint64("subsidy", subsidyReward.Uint64()),
+			log.Uint64("total", totalReward.Uint64()),
+		)
 
-		fTotal, _ := rewardPer.Copy().Mul(weight).Float64()
-		totalReward := uint64(fTotal)
-		fLyr, _ := lyrRewardPer.Copy().Mul(weight).Float64()
-		finalRewards = append(finalRewards, &types.Reward{
-			Layer:       lid,
-			Coinbase:    reward.Coinbase,
-			TotalReward: totalReward,
-			LayerReward: uint64(fLyr),
-		})
+		reward := &types.Reward{
+			Layer:       lctx.Layer,
+			Coinbase:    blockReward.Coinbase,
+			TotalReward: totalReward.Uint64(),
+			LayerReward: subsidyReward.Uint64(),
+		}
+		if err := rewards.Add(tx, reward); err != nil {
+			return fmt.Errorf("%w: %s", core.ErrInternal, err.Error())
+		}
+		account, err := ss.Get(blockReward.Coinbase)
+		if err != nil {
+			return fmt.Errorf("%w: %s", core.ErrInternal, err.Error())
+		}
+		account.Balance += reward.TotalReward
+		if err := ss.Update(account); err != nil {
+			return fmt.Errorf("%w: %s", core.ErrInternal, err.Error())
+		}
+		transferred += totalReward.Uint64()
 	}
-	return finalRewards, nil
+	v.logger.With().Debug("rewards for layer",
+		lctx.Layer,
+		log.Uint32("after genesis", layersAfterEffectiveGenesis),
+		log.Uint64("subsidy estimated", subsidy),
```

Comment:
why the term "estimated" here and for total?

Reply:
actual amount computed from how much was transfered

Reply:
I don't understand - "estimated" here means "before rounding"? or something else?

Reply:
before rounding